src/v/raft/tests/raft_fixture.cc

```diff
@@ -59,8 +59,10 @@ ss::future<> channel::stop() {
 
         auto f = _gate.close();
 
-        for (auto& m : _messages) {
-            m.resp_data.set_exception(ss::abort_requested_exception());
+        while (!_messages.empty()) {
+            auto msg = std::move(_messages.front());
+            _messages.pop_front();
+            msg.resp_data.set_exception(ss::abort_requested_exception());
```

Comment:
> Sometimes it failed on repeated promise set.

what failed exactly? if the double set occurred in this loop, then it doesn't look like this change would fix that scenario. if the double set occurred elsewhere because this loop left the promises in the messages queue, then is there a race with shutdown that waiting on a gate would fix?

(note, i think we do want to clear this like you are doing, i'm just asking for the commit message to explain what the problem is more precisely).

Reply:
> if the double set occurred elsewhere because this loop left the promises in the messages queue, then is there a race with shutdown that waiting on a gate would fix?

Yes that's the theory. I cannot reproduce it though, and I lost the logs recently (not sure how bazel forms the log path but it overwrote it). I'll see if I can fix it with a gate.

Reply:
maybe it could be reproduced with a strategically placed sleep

Reply:
@dotnwat I'd merge it as is unless you have any concerns. I like to dequeue before getting a future because it forms a clear invariant, while rechecking `_as` in the loop is a logic that needs explanation.